src/escrows/DolaFraxPyUSDEscrow.sol

```diff
@@ -0,0 +1,239 @@
+
+    address public market;
+    IERC20 public token;
+    uint256 public stakedBalance;
+    address public beneficiary;
+
+    mapping(address => bool) public allowlist;
+
+    modifier onlyBeneficiary() {
+        if (msg.sender != beneficiary) revert OnlyBeneficiary();
+        _;
+    }
+
+    modifier onlyBeneficiaryOrAllowlist() {
+        if (msg.sender != beneficiary && !allowlist[msg.sender])
+            revert OnlyBeneficiaryOrAllowlist();
+        _;
+    }
+
+    event AllowClaim(address indexed allowedAddress, bool allowed);
+
+    /**
+    @notice Initialize escrow with a token
+    @dev Must be called right after proxy is created.
+    @param _token The IERC20 token representing the governance token
+    @param _beneficiary The beneficiary who cvxCRV is staked on behalf
+    */
+    function initialize(IERC20 _token, address _beneficiary) public {
+        if (market != address(0)) revert AlreadyInitialized();
+        market = msg.sender;
+        token = _token;
+        token.approve(address(booster), type(uint).max);
+        token.approve(address(yearn), type(uint).max);
+        beneficiary = _beneficiary;
+    }
+
+    /**
+    @notice Withdraws the wrapped token from the reward pool and transfers the associated ERC20 token to a recipient.
+    @dev Will first try to pay from the escrow balance, if not enough or any, will try to pay the missing amount from convex, then from yearn if needed
+    @param recipient The address to receive payment from the escrow
+    @param amount The amount of ERC20 token to be transferred.
+    */
+    function pay(address recipient, uint amount) public {
+        if (msg.sender != market) revert OnlyMarket();
+        uint256 tokenBal = token.balanceOf(address(this));
+        uint256 missingAmount = amount > tokenBal ? amount - tokenBal : 0;
+        // If there are enough tokens in the escrow, transfer the amount
+        if (tokenBal >= amount) {
+            token.safeTransfer(recipient, amount);
+        } else if (stakedBalance >= missingAmount) {
+            // If there are enough staked tokens in convex, withdraw the amount from convex
+            stakedBalance -= missingAmount;
+            rewardPool.withdrawAndUnwrap(missingAmount, false);
+            token.safeTransfer(recipient, amount);
+        } else if (
+            YearnVaultV2Helper.collateralToAsset(
+                yearn,
+                yearn.balanceOf(address(this))
+            ) >= missingAmount
+        ) {
+            // If there are enough tokens in Yearn, withdraw the amount from Yearn
+            uint256 withdrawAmount = YearnVaultV2Helper.assetToCollateral(
+                yearn,
+                missingAmount
+            );
+            // Withdraw from Yearn
+            uint256 assetAmount = yearn.withdraw(withdrawAmount, address(this));
+            // Transfer the amount to the recipient
+            token.safeTransfer(recipient, assetAmount + tokenBal);
+        }
```

Comment:
This should have terminating logic that reverts on no transfer. It's very dangerous for the pay function to be able to execute without transferring anything and without reverting.